hera_cal/io.py

```diff
@@ -864,6 +864,147 @@ def iterate_over_times(self, Nints=1, times=None):
+    inds = {}
+    # Read file metadata to size up arrays and sort times
+    for filename in filenames:
+        if verbose:
+            print(f'Reading header of {filename}')
+        with h5py.File(filename, 'r') as f:
+            h = f['/Header']
+            if check:
+                assert int(h['Nspws'][()]) == 1  # not a hera file
+            if len(times) == 0:
+                info['freqs'] = h['freq_array'][()]
+                pol_array = h['polarization_array'][()]
+                npols = pol_array.size
+                pol_indices = {POL_NUM2STR_DICT[n]: cnt for cnt, n in enumerate(pol_array)}
+                info['pols'] = list(pol_indices.keys())
+            elif check:
+                assert int(h['Nfreqs'][()]) == info['freqs'].size
+            ntimes = int(h['Ntimes'][()])
+            times.append((h['time_array'][:ntimes], filename))
+            ant1_array = h['ant_1_array'][::ntimes]
+            ant2_array = h['ant_2_array'][::ntimes]
+            _hash = hash((ant1_array.tobytes(), ant2_array.tobytes()))
+            # map baselines to array indices for each unique antenna order
+            if _hash not in inds:
+                inds[_hash] = {(i, j): slice(n * ntimes, (n + 1) * ntimes)
+                               for n, (i, j) in enumerate(zip(ant1_array,
+                                                              ant2_array))}
+                if 'bls' not in info:
+                    info['bls'] = set(inds[_hash].keys())
+                else:
+                    info['bls'].intersection_update(set(inds[_hash].keys()))
+            bl2ind[filename] = inds[_hash]
+
+    if bls is None:
+        # generate a set of bls if we didn't have one passed in
+        if pols is None:
+            pols = list(pol_indices.keys())
+        bls = info['bls']
+        bls = set(bl + (p,) for bl in bls for p in pols)
+    else:
+        # if length 2 baselines are passed in, add on polarizations
+        bls_len2 = set(bl for bl in bls if len(bl) == 2)
+        if len(bls_len2) > 0:
+            if pols is None:
+                pols = list(pol_indices.keys())
+            bls = set(bl for bl in bls if len(bl) == 3)
+            bls = bls.union([bl + (p,) for bl in bls_len2 for p in pols])
+        # record polarizations as total of ones indexed in bls
+        pols = set(bl[2] for bl in bls)
+    # sort files by time of first integration
+    times.sort(key=lambda x: x[0][0])
+    filenames = (v[-1] for v in times)
+    times = np.concatenate([t[0] for t in times], axis=0)
+    info['times'] = times
+
+    # preallocate buffers
+    rv = {}
+    if read_data:
+        rv['visdata'] = {bl: np.empty((times.size, info['freqs'].size), dtype=dtype) for bl in bls}
+    if read_flags:
+        rv['flags'] = {bl: np.empty((times.size, info['freqs'].size), dtype=bool) for bl in bls}
+    if read_nsamples:
+        rv['nsamples'] = {bl: np.empty((times.size, info['freqs'].size), dtype=np.float32) for bl in bls}
```

Comment:
Do you want to cast these as DataContainers? I feel like that would make this more interoperable with the rest of `hera_cal` without adding much overhead.

Reply:
Also, as DataContainers know to expect metadata as class attributes, namely `'ants', 'data_ants', 'antpos', 'data_antpos', 'freqs', 'times', 'lsts', 'times_by_bl', 'lsts_by_bl'`, so perhaps a lot of the metadata currently stored in the `info` object could be stored instead in these objects. There might be some duplication of metadata as a result, but I think that'd be trivial in terms of extra memory and runtime, no?

Reply:
I don't think it's correct to think of this function as a stand-in for DataContainers or any other hera_cal object. It should be thought of as a stand-in for pyuvdata. The goal should be to get the data into a generic format (dicts are fine) as fast as possible. Then HERAData can take care of the rest.

That said, I agree that "info" is light on the metadata it reads out and should be expanded as far as it can without hurting performance. Is this a complete list of commonly used file metadata?

Reply:
I added antpos, but lst is not in the files, so i'm not computing it

Reply:
Ok, then I think perhaps the next thing to do (and this is probably on me) is to write a wrapper around this function that puts things into a format that might be closer to a drop-in replacement for HERAData in the code.